File: compat/suites/python-sdk/groups/elasticache.py
"""
groups/elasticache.py — ElastiCache compatibility test implementations for the Python suite.

Groups:
  elasticache-clusters           — Cache cluster lifecycle (create, describe, delete)
  elasticache-modify             — Modify cache clusters and replication groups
  elasticache-replication-groups — Replication group lifecycle
  elasticache-subnet-groups      — Cache subnet group lifecycle
  elasticache-parameter-groups   — Cache parameter group lifecycle
"""

from __future__ import annotations
from lib.harness import TestContext
from lib.clients import make_clients


def _elasticache(ctx: TestContext):
    return make_clients(ctx.endpoint, ctx.region).elasticache


# ── elasticache-clusters ──────────────────────────────────────────────────────

def teardown_elasticache_clusters(ctx: TestContext) -> None:
    ec = _elasticache(ctx)
    for key in ("_clusterId", "_describeClusterId", "_deleteClusterId"):
        cid = ctx.get(key)
        if cid:
            try:
                ec.delete_cache_cluster(CacheClusterId=cid)
            except Exception:
                pass


def CreateCacheCluster(ctx: TestContext) -> None:
    ec = _elasticache(ctx)
    cid = f"compat-cluster-{ctx.run_id}"
    ctx["_clusterId"] = cid
    resp = ec.create_cache_cluster(
        CacheClusterId=cid,
        Engine="redis",
        CacheNodeType="cache.t3.micro",
        NumCacheNodes=1,
    )
    cluster = resp.get("CacheCluster", {})
    if cluster.get("CacheClusterId") != cid:
        raise AssertionError(f"CreateCacheCluster: expected id {cid}, got {cluster.get('CacheClusterId')}")
    status = cluster.get("CacheClusterStatus")
    if status not in ("creating", "available"):
        raise AssertionError(f"CreateCacheCluster: unexpected status {status}")


def DescribeCacheClusters(ctx: TestContext) -> None:
    ec = _elasticache(ctx)
    cid = f"compat-describe-{ctx.run_id}"
    ctx["_describeClusterId"] = cid
    ec.create_cache_cluster(
        CacheClusterId=cid,
        Engine="redis",
        CacheNodeType="cache.t3.micro",
        NumCacheNodes=1,
    )
    resp = ec.describe_cache_clusters(CacheClusterId=cid)
    clusters = resp.get("CacheClusters", [])
    if len(clusters) != 1:
        raise AssertionError(f"DescribeCacheClusters: expected 1 cluster, got {len(clusters)}")
    if clusters[0].get("CacheClusterId") != cid:
        raise AssertionError(f"DescribeCacheClusters: expected id {cid}, got {clusters[0].get('CacheClusterId')}")


def DeleteCacheCluster(ctx: TestContext) -> None:
    ec = _elasticache(ctx)
    cid = f"compat-delete-{ctx.run_id}"
    ctx["_deleteClusterId"] = cid
    ec.create_cache_cluster(
        CacheClusterId=cid,
        Engine="redis",
        CacheNodeType="cache.t3.micro",
        NumCacheNodes=1,
    )
    resp = ec.delete_cache_cluster(CacheClusterId=cid)
    cluster = resp.get("CacheCluster", {})
    if cluster.get("CacheClusterStatus") != "deleting":
        raise AssertionError(f"DeleteCacheCluster: expected status deleting, got {cluster.get('CacheClusterStatus')}")


# ── elasticache-modify ────────────────────────────────────────────────────────

def teardown_elasticache_modify(ctx: TestContext) -> None:
    ec = _elasticache(ctx)
    for key in ("_modifyClusterId", "_modifyRgId"):
        rid = ctx.get(key)
        if not rid:
            continue
        try:
            if key == "_modifyClusterId":
                ec.delete_cache_cluster(CacheClusterId=rid)
            else:
                ec.delete_replication_group(ReplicationGroupId=rid)
        except Exception:
            pass


def ModifyCacheCluster(ctx: TestContext) -> None:
    ec = _elasticache(ctx)
    cid = f"compat-modify-cluster-{ctx.run_id}"
    ctx["_modifyClusterId"] = cid
    ec.create_cache_cluster(
        CacheClusterId=cid,
        Engine="redis",
        CacheNodeType="cache.t3.micro",
        NumCacheNodes=1,
    )
    resp = ec.modify_cache_cluster(
        CacheClusterId=cid,
        CacheNodeType="cache.t3.small",
        ApplyImmediately=True,
    )
    cluster = resp.get("CacheCluster", {})
    if cluster.get("CacheClusterId") != cid:
        raise AssertionError(f"ModifyCacheCluster: expected id {cid}, got {cluster.get('CacheClusterId')}")
    if cluster.get("CacheClusterStatus") != "modifying":
        raise AssertionError(f"ModifyCacheCluster: expected status modifying, got {cluster.get('CacheClusterStatus')}")
    if cluster.get("CacheNodeType") != "cache.t3.small":
        raise AssertionError(f"ModifyCacheCluster: expected type cache.t3.small, got {cluster.get('CacheNodeType')}")


def ModifyReplicationGroup(ctx: TestContext) -> None:
    ec = _elasticache(ctx)
    rgid = f"compat-modify-rg-{ctx.run_id}"
    ctx["_modifyRgId"] = rgid
    ec.create_replication_group(
        ReplicationGroupId=rgid,
        ReplicationGroupDescription="original",
        CacheNodeType="cache.t3.micro",
    )
    resp = ec.modify_replication_group(
        ReplicationGroupId=rgid,
        ReplicationGroupDescription="updated",
        ApplyImmediately=True,
    )
    rg = resp.get("ReplicationGroup", {})
    if rg.get("ReplicationGroupId") != rgid:
        raise AssertionError(f"ModifyReplicationGroup: expected id {rgid}, got {rg.get('ReplicationGroupId')}")
    if rg.get("Status") != "modifying":
        raise AssertionError(f"ModifyReplicationGroup: expected status modifying, got {rg.get('Status')}")
    if rg.get("Description") != "updated":
        raise AssertionError(f"ModifyReplicationGroup: expected desc updated, got {rg.get('Description')}")


# ── elasticache-replication-groups ────────────────────────────────────────────

def teardown_elasticache_replication_groups(ctx: TestContext) -> None:
    ec = _elasticache(ctx)
    for key in ("_rgId", "_rgDescId"):
        rgid = ctx.get(key)
        if rgid:
            try:
                ec.delete_replication_group(ReplicationGroupId=rgid)
            except Exception:
                pass


def CreateReplicationGroup(ctx: TestContext) -> None:
    ec = _elasticache(ctx)
    rgid = f"compat-rg-{ctx.run_id}"
    ctx["_rgId"] = rgid
    resp = ec.create_replication_group(
        ReplicationGroupId=rgid,
        ReplicationGroupDescription="compat test group",
        CacheNodeType="cache.t3.micro",
    )
    rg = resp.get("ReplicationGroup", {})
    if rg.get("ReplicationGroupId") != rgid:
        raise AssertionError(f"CreateReplicationGroup: expected id {rgid}, got {rg.get('ReplicationGroupId')}")
    status = rg.get("Status")
    if status not in ("creating", "available"):
        raise AssertionError(f"CreateReplicationGroup: unexpected status {status}")


def DescribeReplicationGroups(ctx: TestContext) -> None:
    ec = _elasticache(ctx)
    rgid = f"compat-rg-desc-{ctx.run_id}"
    ctx["_rgDescId"] = rgid
    ec.create_replication_group(
        ReplicationGroupId=rgid,
        ReplicationGroupDescription="compat describe",
        CacheNodeType="cache.t3.micro",
    )
    resp = ec.describe_replication_groups(ReplicationGroupId=rgid)
    groups = resp.get("ReplicationGroups", [])
    if len(groups) != 1:
        raise AssertionError(f"DescribeReplicationGroups: expected 1 group, got {len(groups)}")
    if groups[0].get("ReplicationGroupId") != rgid:
        raise AssertionError(f"DescribeReplicationGroups: expected id {rgid}, got {groups[0].get('ReplicationGroupId')}")


# ── elasticache-subnet-groups ─────────────────────────────────────────────────

def teardown_elasticache_subnet_groups(ctx: TestContext) -> None:
    ec = _elasticache(ctx)
    for key in ("_sngrpName", "_sngrpDescName"):
        name = ctx.get(key)
        if name:
            try:
                ec.delete_cache_subnet_group(CacheSubnetGroupName=name)
            except Exception:
                pass


def CreateCacheSubnetGroup(ctx: TestContext) -> None:
    ec = _elasticache(ctx)
    name = f"compat-sngrp-{ctx.run_id}"
    ctx["_sngrpName"] = name
    resp = ec.create_cache_subnet_group(
        CacheSubnetGroupName=name,
        CacheSubnetGroupDescription="compat test",
        SubnetIds=["subnet-aabbccdd"],
    )
    sngrp = resp.get("CacheSubnetGroup", {})
    if sngrp.get("CacheSubnetGroupName") != name:
        raise AssertionError(f"CreateCacheSubnetGroup: expected name {name}, got {sngrp.get('CacheSubnetGroupName')}")


def DescribeCacheSubnetGroups(ctx: TestContext) -> None:
    ec = _elasticache(ctx)
    name = f"compat-sngrp-desc-{ctx.run_id}"
    ctx["_sngrpDescName"] = name
    ec.create_cache_subnet_group(
        CacheSubnetGroupName=name,
        CacheSubnetGroupDescription="describe test",
        SubnetIds=["subnet-11223344"],
    )
    resp = ec.describe_cache_subnet_groups(CacheSubnetGroupName=name)
    groups = resp.get("CacheSubnetGroups", [])
    if len(groups) != 1:
        raise AssertionError(f"DescribeCacheSubnetGroups: expected 1 group, got {len(groups)}")
    if groups[0].get("CacheSubnetGroupName") != name:
        raise AssertionError(f"DescribeCacheSubnetGroups: expected name {name}, got {groups[0].get('CacheSubnetGroupName')}")


# ── elasticache-parameter-groups ──────────────────────────────────────────────

def teardown_elasticache_parameter_groups(ctx: TestContext) -> None:
    ec = _elasticache(ctx)
    for key in ("_pgName", "_pgDescName", "_pgDelName"):
        name = ctx.get(key)
        if name:
            try:
                ec.delete_cache_parameter_group(CacheParameterGroupName=name)
            except Exception:
                pass


def CreateCacheParameterGroup(ctx: TestContext) -> None:
    ec = _elasticache(ctx)
    name = f"compat-pg-{ctx.run_id}"
    ctx["_pgName"] = name
    resp = ec.create_cache_parameter_group(
        CacheParameterGroupName=name,
        CacheParameterGroupFamily="redis7",
        Description="compat test group",
    )
    pg = resp.get("CacheParameterGroup", {})
    if pg.get("CacheParameterGroupName") != name:
        raise AssertionError(f"CreateCacheParameterGroup: expected name {name}, got {pg.get('CacheParameterGroupName')}")
    if pg.get("CacheParameterGroupFamily") != "redis7":
        raise AssertionError(f"CreateCacheParameterGroup: expected family redis7, got {pg.get('CacheParameterGroupFamily')}")
    arn = pg.get("ARN", "")
    if name not in arn:
        raise AssertionError(f"CreateCacheParameterGroup: ARN {arn!r} does not contain {name}")


def DescribeCacheParameterGroups(ctx: TestContext) -> None:
    ec = _elasticache(ctx)
    name = f"compat-pg-desc-{ctx.run_id}"
    ctx["_pgDescName"] = name
    ec.create_cache_parameter_group(
        CacheParameterGroupName=name,
        CacheParameterGroupFamily="redis7",
        Description="describe test",
    )
    resp = ec.describe_cache_parameter_groups(CacheParameterGroupName=name)
    groups = resp.get("CacheParameterGroups", [])
    if len(groups) != 1:
        raise AssertionError(f"DescribeCacheParameterGroups: expected 1 group, got {len(groups)}")
    if groups[0].get("CacheParameterGroupName") != name:
        raise AssertionError(f"DescribeCacheParameterGroups: expected name {name}, got {groups[0].get('CacheParameterGroupName')}")
    if groups[0].get("CacheParameterGroupFamily") != "redis7":
        raise AssertionError(f"DescribeCacheParameterGroups: expected family redis7, got {groups[0].get('CacheParameterGroupFamily')}")


def DeleteCacheParameterGroup(ctx: TestContext) -> None:
    ec = _elasticache(ctx)
    name = f"compat-pg-del-{ctx.run_id}"
    ctx["_pgDelName"] = name
    ec.create_cache_parameter_group(
        CacheParameterGroupName=name,
        CacheParameterGroupFamily="redis7",
        Description="delete test",
    )
    ec.delete_cache_parameter_group(CacheParameterGroupName=name)
    try:
        ec.describe_cache_parameter_groups(CacheParameterGroupName=name)
        raise AssertionError("DeleteCacheParameterGroup: group still exists after deletion")
    except Exception as e:
        if "CacheParameterGroupNotFound" not in str(e):
            raise AssertionError(f"DeleteCacheParameterGroup: unexpected error: {e}")


# ── ImplMap ───────────────────────────────────────────────────────────────────

IMPLS = {
    "CreateCacheCluster": CreateCacheCluster,
    "DescribeCacheClusters": DescribeCacheClusters,
    "DeleteCacheCluster": DeleteCacheCluster,
    "ModifyCacheCluster": ModifyCacheCluster,
    "ModifyReplicationGroup": ModifyReplicationGroup,
    "CreateReplicationGroup": CreateReplicationGroup,
    "DescribeReplicationGroups": DescribeReplicationGroups,
    "CreateCacheSubnetGroup": CreateCacheSubnetGroup,
    "DescribeCacheSubnetGroups": DescribeCacheSubnetGroups,
    "CreateCacheParameterGroup": CreateCacheParameterGroup,
    "DescribeCacheParameterGroups": DescribeCacheParameterGroups,
    "DeleteCacheParameterGroup": DeleteCacheParameterGroup,
}

SETUP = {}

TEARDOWN = {
    "elasticache-clusters": teardown_elasticache_clusters,
    "elasticache-modify": teardown_elasticache_modify,
    "elasticache-replication-groups": teardown_elasticache_replication_groups,
    "elasticache-subnet-groups": teardown_elasticache_subnet_groups,
    "elasticache-parameter-groups": teardown_elasticache_parameter_groups,
}

File: compat/suites/go-sdk/internal/groups/elasticache.go
package groups

import (
	"context"
	"fmt"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/elasticache"
	"github.com/overcast-sh/overcast-compat-go-sdk/internal/clients"
	"github.com/overcast-sh/overcast-compat-go-sdk/internal/harness"
)

func ElastiCache(c *clients.Clients) ServiceGroup {
	g := &elasticacheGroup{c: c}
	return ServiceGroup{
		Impls: map[string]harness.TestFn{
			"CreateCacheCluster":           g.CreateCacheCluster,
			"DescribeCacheClusters":        g.DescribeCacheClusters,
			"DeleteCacheCluster":           g.DeleteCacheCluster,
			"ModifyCacheCluster":           g.ModifyCacheCluster,
			"ModifyReplicationGroup":       g.ModifyReplicationGroup,
			"CreateReplicationGroup":       g.CreateReplicationGroup,
			"DescribeReplicationGroups":    g.DescribeReplicationGroups,
			"CreateCacheSubnetGroup":       g.CreateCacheSubnetGroup,
			"DescribeCacheSubnetGroups":    g.DescribeCacheSubnetGroups,
			"CreateCacheParameterGroup":    g.CreateCacheParameterGroup,
			"DescribeCacheParameterGroups": g.DescribeCacheParameterGroups,
			"DeleteCacheParameterGroup":    g.DeleteCacheParameterGroup,
		},
		Setup: map[string]func(context.Context, *harness.TestContext) error{
			"elasticache-clusters":           g.setupNoop,
			"elasticache-modify":             g.setupNoop,
			"elasticache-replication-groups": g.setupNoop,
			"elasticache-subnet-groups":      g.setupNoop,
			"elasticache-parameter-groups":   g.setupNoop,
		},
		Teardown: map[string]func(context.Context, *harness.TestContext) error{
			"elasticache-clusters":           g.teardownClusters,
			"elasticache-modify":             g.teardownModify,
			"elasticache-replication-groups": g.teardownReplicationGroups,
			"elasticache-subnet-groups":      g.teardownSubnetGroups,
			"elasticache-parameter-groups":   g.teardownParameterGroups,
		},
	}
}

type elasticacheGroup struct{ c *clients.Clients }

func (g *elasticacheGroup) cl() *elasticache.Client { return g.c.ElastiCache() }
func (g *elasticacheGroup) setupNoop(_ context.Context, _ *harness.TestContext) error {
	return nil
}
func (g *elasticacheGroup) uniqueName(desc string) string {
	return fmt.Sprintf("compat-%s-%d", desc, time.Now().Unix())
}

// ─── elasticache-clusters ─────────────────────────────────────────────────────

func (g *elasticacheGroup) CreateCacheCluster(ctx context.Context, t *harness.TestContext) error {
	id := g.uniqueName("cluster")
	t.Set("_clusterId", id)
	_, err := g.cl().CreateCacheCluster(ctx, &elasticache.CreateCacheClusterInput{
		CacheClusterId: aws.String(id),
		Engine:         aws.String("redis"),
		CacheNodeType:  aws.String("cache.t3.micro"),
		NumCacheNodes:  aws.Int32(1),
	})
	if err != nil {
		return err
	}
	resp, err := g.cl().DescribeCacheClusters(ctx, &elasticache.DescribeCacheClustersInput{
		CacheClusterId: aws.String(id),
	})
	if err != nil {
		return fmt.Errorf("elasticache CreateCacheCluster: describe failed: %w", err)
	}
	if len(resp.CacheClusters) != 1 {
		return fmt.Errorf("elasticache CreateCacheCluster: expected 1 cluster, got %d", len(resp.CacheClusters))
	}
	if aws.ToString(resp.CacheClusters[0].CacheClusterId) != id {
		return fmt.Errorf("elasticache CreateCacheCluster: expected CacheClusterId %q, got %v", id, aws.ToString(resp.CacheClusters[0].CacheClusterId))
	}
	return nil
}

func (g *elasticacheGroup) DescribeCacheClusters(ctx context.Context, t *harness.TestContext) error {
	id := g.uniqueName("describe")
	t.Set("_describeClusterId", id)
	_, err := g.cl().CreateCacheCluster(ctx, &elasticache.CreateCacheClusterInput{
		CacheClusterId: aws.String(id),
		Engine:         aws.String("redis"),
		CacheNodeType:  aws.String("cache.t3.micro"),
		NumCacheNodes:  aws.Int32(1),
	})
	if err != nil {
		return err
	}
	resp, err := g.cl().DescribeCacheClusters(ctx, &elasticache.DescribeCacheClustersInput{
		CacheClusterId: aws.String(id),
	})
	if err != nil {
		return fmt.Errorf("elasticache DescribeCacheClusters: describe failed: %w", err)
	}
	if len(resp.CacheClusters) != 1 {
		return fmt.Errorf("elasticache DescribeCacheClusters: expected 1 cluster, got %d", len(resp.CacheClusters))
	}
	if aws.ToString(resp.CacheClusters[0].CacheClusterId) != id {
		return fmt.Errorf("elasticache DescribeCacheClusters: expected CacheClusterId %q, got %v", id, aws.ToString(resp.CacheClusters[0].CacheClusterId))
	}
	return nil
}

func (g *elasticacheGroup) DeleteCacheCluster(ctx context.Context, t *harness.TestContext) error {
	id := g.uniqueName("delete")
	t.Set("_deleteClusterId", id)
	_, err := g.cl().CreateCacheCluster(ctx, &elasticache.CreateCacheClusterInput{
		CacheClusterId: aws.String(id),
		Engine:         aws.String("redis"),
		CacheNodeType:  aws.String("cache.t3.micro"),
		NumCacheNodes:  aws.Int32(1),
	})
	if err != nil {
		return err
	}
	resp, err := g.cl().DeleteCacheCluster(ctx, &elasticache.DeleteCacheClusterInput{
		CacheClusterId: aws.String(id),
	})
	if err != nil {
		return fmt.Errorf("elasticache DeleteCacheCluster: delete failed: %w", err)
	}
	if resp.CacheCluster == nil {
		return fmt.Errorf("elasticache DeleteCacheCluster: no CacheCluster in response")
	}
	if aws.ToString(resp.CacheCluster.CacheClusterStatus) != "deleting" {
		return fmt.Errorf("elasticache DeleteCacheCluster: expected CacheClusterStatus 'deleting', got %v", aws.ToString(resp.CacheCluster.CacheClusterStatus))
	}
	return nil
}

func (g *elasticacheGroup) teardownClusters(ctx context.Context, t *harness.TestContext) error {
	for _, key := range []string{"_clusterId", "_describeClusterId", "_deleteClusterId"} {
		if id := t.GetString(key); id != "" {
			g.cl().DeleteCacheCluster(ctx, &elasticache.DeleteCacheClusterInput{CacheClusterId: aws.String(id)}) //nolint:errcheck
		}
	}
	return nil
}

// ─── elasticache-modify ───────────────────────────────────────────────────────

func (g *elasticacheGroup) ModifyCacheCluster(ctx context.Context, t *harness.TestContext) error {
	id := g.uniqueName("modify-cluster")
	t.Set("_modifyClusterId", id)
	_, err := g.cl().CreateCacheCluster(ctx, &elasticache.CreateCacheClusterInput{
		CacheClusterId: aws.String(id),
		Engine:         aws.String("redis"),
		CacheNodeType:  aws.String("cache.t3.micro"),
		NumCacheNodes:  aws.Int32(1),
	})
	if err != nil {
		return err
	}
	resp, err := g.cl().ModifyCacheCluster(ctx, &elasticache.ModifyCacheClusterInput{
		CacheClusterId:   aws.String(id),
		CacheNodeType:    aws.String("cache.t3.small"),
		ApplyImmediately: aws.Bool(true),
	})
	if err != nil {
		return fmt.Errorf("elasticache ModifyCacheCluster: modify failed: %w", err)
	}
	if resp.CacheCluster == nil {
		return fmt.Errorf("elasticache ModifyCacheCluster: no CacheCluster in response")
	}
	if aws.ToString(resp.CacheCluster.CacheClusterId) != id {
		return fmt.Errorf("elasticache ModifyCacheCluster: expected CacheClusterId %q, got %v", id, aws.ToString(resp.CacheCluster.CacheClusterId))
	}
	if aws.ToString(resp.CacheCluster.CacheClusterStatus) != "modifying" {
		return fmt.Errorf("elasticache ModifyCacheCluster: expected CacheClusterStatus 'modifying', got %v", aws.ToString(resp.CacheCluster.CacheClusterStatus))
	}
	if aws.ToString(resp.CacheCluster.CacheNodeType) != "cache.t3.small" {
		return fmt.Errorf("elasticache ModifyCacheCluster: expected CacheNodeType 'cache.t3.small', got %v", aws.ToString(resp.CacheCluster.CacheNodeType))
	}
	return nil
}

func (g *elasticacheGroup) ModifyReplicationGroup(ctx context.Context, t *harness.TestContext) error {
	id := g.uniqueName("modify-rg")
	t.Set("_modifyRgId", id)
	_, err := g.cl().CreateReplicationGroup(ctx, &elasticache.CreateReplicationGroupInput{
		ReplicationGroupId:          aws.String(id),
		ReplicationGroupDescription: aws.String("original"),
		CacheNodeType:               aws.String("cache.t3.micro"),
	})
	if err != nil {
		return err
	}
	resp, err := g.cl().ModifyReplicationGroup(ctx, &elasticache.ModifyReplicationGroupInput{
		ReplicationGroupId:          aws.String(id),
		ReplicationGroupDescription: aws.String("updated"),
		ApplyImmediately:            aws.Bool(true),
	})
	if err != nil {
		return fmt.Errorf("elasticache ModifyReplicationGroup: modify failed: %w", err)
	}
	if resp.ReplicationGroup == nil {
		return fmt.Errorf("elasticache ModifyReplicationGroup: no ReplicationGroup in response")
	}
	if aws.ToString(resp.ReplicationGroup.ReplicationGroupId) != id {
		return fmt.Errorf("elasticache ModifyReplicationGroup: expected ReplicationGroupId %q, got %v", id, aws.ToString(resp.ReplicationGroup.ReplicationGroupId))
	}
	if aws.ToString(resp.ReplicationGroup.Status) != "modifying" {
		return fmt.Errorf("elasticache ModifyReplicationGroup: expected Status 'modifying', got %v", aws.ToString(resp.ReplicationGroup.Status))
	}
	if aws.ToString(resp.ReplicationGroup.Description) != "updated" {
		return fmt.Errorf("elasticache ModifyReplicationGroup: expected Description 'updated', got %v", aws.ToString(resp.ReplicationGroup.Description))
	}
	return nil
}

func (g *elasticacheGroup) teardownModify(ctx context.Context, t *harness.TestContext) error {
	if id := t.GetString("_modifyClusterId"); id != "" {
		g.cl().DeleteCacheCluster(ctx, &elasticache.DeleteCacheClusterInput{CacheClusterId: aws.String(id)}) //nolint:errcheck
	}
	if id := t.GetString("_modifyRgId"); id != "" {
		g.cl().DeleteReplicationGroup(ctx, &elasticache.DeleteReplicationGroupInput{ReplicationGroupId: aws.String(id)}) //nolint:errcheck
	}
	return nil
}

// ─── elasticache-replication-groups ───────────────────────────────────────────

func (g *elasticacheGroup) CreateReplicationGroup(ctx context.Context, t *harness.TestContext) error {
	id := g.uniqueName("rg")
	t.Set("_rgId", id)
	_, err := g.cl().CreateReplicationGroup(ctx, &elasticache.CreateReplicationGroupInput{
		ReplicationGroupId:          aws.String(id),
		ReplicationGroupDescription: aws.String("compat test group"),
		CacheNodeType:               aws.String("cache.t3.micro"),
	})
	if err != nil {
		return err
	}
	resp, err := g.cl().DescribeReplicationGroups(ctx, &elasticache.DescribeReplicationGroupsInput{
		ReplicationGroupId: aws.String(id),
	})
	if err != nil {
		return fmt.Errorf("elasticache CreateReplicationGroup: describe failed: %w", err)
	}
	if len(resp.ReplicationGroups) != 1 {
		return fmt.Errorf("elasticache CreateReplicationGroup: expected 1 group, got %d", len(resp.ReplicationGroups))
	}
	if aws.ToString(resp.ReplicationGroups[0].ReplicationGroupId) != id {
		return fmt.Errorf("elasticache CreateReplicationGroup: expected ReplicationGroupId %q, got %v", id, aws.ToString(resp.ReplicationGroups[0].ReplicationGroupId))
	}
	return nil
}

func (g *elasticacheGroup) DescribeReplicationGroups(ctx context.Context, t *harness.TestContext) error {
	id := g.uniqueName("rg-desc")
	t.Set("_rgDescId", id)
	_, err := g.cl().CreateReplicationGroup(ctx, &elasticache.CreateReplicationGroupInput{
		ReplicationGroupId:          aws.String(id),
		ReplicationGroupDescription: aws.String("compat describe"),
		CacheNodeType:               aws.String("cache.t3.micro"),
	})
	if err != nil {
		return err
	}
	resp, err := g.cl().DescribeReplicationGroups(ctx, &elasticache.DescribeReplicationGroupsInput{
		ReplicationGroupId: aws.String(id),
	})
	if err != nil {
		return fmt.Errorf("elasticache DescribeReplicationGroups: describe failed: %w", err)
	}
	if len(resp.ReplicationGroups) != 1 {
		return fmt.Errorf("elasticache DescribeReplicationGroups: expected 1 group, got %d", len(resp.ReplicationGroups))
	}
	if aws.ToString(resp.ReplicationGroups[0].ReplicationGroupId) != id {
		return fmt.Errorf("elasticache DescribeReplicationGroups: expected ReplicationGroupId %q, got %v", id, aws.ToString(resp.ReplicationGroups[0].ReplicationGroupId))
	}
	return nil
}

func (g *elasticacheGroup) teardownReplicationGroups(ctx context.Context, t *harness.TestContext) error {
	for _, key := range []string{"_rgId", "_rgDescId"} {
		if id := t.GetString(key); id != "" {
			g.cl().DeleteReplicationGroup(ctx, &elasticache.DeleteReplicationGroupInput{ReplicationGroupId: aws.String(id)}) //nolint:errcheck
		}
	}
	return nil
}

// ─── elasticache-subnet-groups ────────────────────────────────────────────────

func (g *elasticacheGroup) CreateCacheSubnetGroup(ctx context.Context, t *harness.TestContext) error {
	name := g.uniqueName("sngrp")
	t.Set("_sngrpName", name)
	resp, err := g.cl().CreateCacheSubnetGroup(ctx, &elasticache.CreateCacheSubnetGroupInput{
		CacheSubnetGroupName:        aws.String(name),
		CacheSubnetGroupDescription: aws.String("compat test"),
		SubnetIds:                   []string{"subnet-aabbccdd"},
	})
	if err != nil {
		return fmt.Errorf("elasticache CreateCacheSubnetGroup: create failed: %w", err)
	}
	if resp.CacheSubnetGroup == nil {
		return fmt.Errorf("elasticache CreateCacheSubnetGroup: no CacheSubnetGroup in response")
	}
	if aws.ToString(resp.CacheSubnetGroup.CacheSubnetGroupName) != name {
		return fmt.Errorf("elasticache CreateCacheSubnetGroup: expected CacheSubnetGroupName %q, got %v", name, aws.ToString(resp.CacheSubnetGroup.CacheSubnetGroupName))
	}
	return nil
}

func (g *elasticacheGroup) DescribeCacheSubnetGroups(ctx context.Context, t *harness.TestContext) error {
	name := g.uniqueName("sngrp-desc")
	t.Set("_sngrpDescName", name)
	_, err := g.cl().CreateCacheSubnetGroup(ctx, &elasticache.CreateCacheSubnetGroupInput{
		CacheSubnetGroupName:        aws.String(name),
		CacheSubnetGroupDescription: aws.String("describe test"),
		SubnetIds:                   []string{"subnet-11223344"},
	})
	if err != nil {
		return err
	}
	resp, err := g.cl().DescribeCacheSubnetGroups(ctx, &elasticache.DescribeCacheSubnetGroupsInput{
		CacheSubnetGroupName: aws.String(name),
	})
	if err != nil {
		return fmt.Errorf("elasticache DescribeCacheSubnetGroups: describe failed: %w", err)
	}
	if len(resp.CacheSubnetGroups) != 1 {
		return fmt.Errorf("elasticache DescribeCacheSubnetGroups: expected 1 group, got %d", len(resp.CacheSubnetGroups))
	}
	if aws.ToString(resp.CacheSubnetGroups[0].CacheSubnetGroupName) != name {
		return fmt.Errorf("elasticache DescribeCacheSubnetGroups: expected CacheSubnetGroupName %q, got %v", name, aws.ToString(resp.CacheSubnetGroups[0].CacheSubnetGroupName))
	}
	return nil
}

func (g *elasticacheGroup) teardownSubnetGroups(ctx context.Context, t *harness.TestContext) error {
	for _, key := range []string{"_sngrpName", "_sngrpDescName"} {
		if name := t.GetString(key); name != "" {
			g.cl().DeleteCacheSubnetGroup(ctx, &elasticache.DeleteCacheSubnetGroupInput{CacheSubnetGroupName: aws.String(name)}) //nolint:errcheck
		}
	}
	return nil
}

// ─── elasticache-parameter-groups ─────────────────────────────────────────────

func (g *elasticacheGroup) CreateCacheParameterGroup(ctx context.Context, t *harness.TestContext) error {
	name := g.uniqueName("pg")
	t.Set("_pgName", name)
	resp, err := g.cl().CreateCacheParameterGroup(ctx, &elasticache.CreateCacheParameterGroupInput{
		CacheParameterGroupName:   aws.String(name),
		CacheParameterGroupFamily: aws.String("redis7"),
		Description:               aws.String("compat test group"),
	})
	if err != nil {
		return fmt.Errorf("elasticache CreateCacheParameterGroup: create failed: %w", err)
	}
	if resp.CacheParameterGroup == nil {
		return fmt.Errorf("elasticache CreateCacheParameterGroup: no CacheParameterGroup in response")
	}
	if aws.ToString(resp.CacheParameterGroup.CacheParameterGroupName) != name {
		return fmt.Errorf("elasticache CreateCacheParameterGroup: expected CacheParameterGroupName %q, got %v", name, aws.ToString(resp.CacheParameterGroup.CacheParameterGroupName))
	}
	if aws.ToString(resp.CacheParameterGroup.CacheParameterGroupFamily) != "redis7" {
		return fmt.Errorf("elasticache CreateCacheParameterGroup: expected CacheParameterGroupFamily 'redis7', got %v", aws.ToString(resp.CacheParameterGroup.CacheParameterGroupFamily))
	}
	if arn := aws.ToString(resp.CacheParameterGroup.ARN); arn == "" {
		return fmt.Errorf("elasticache CreateCacheParameterGroup: ARN is empty")
	}
	return nil
}

func (g *elasticacheGroup) DescribeCacheParameterGroups(ctx context.Context, t *harness.TestContext) error {
	name := g.uniqueName("pg-desc")
	t.Set("_pgDescName", name)
	_, err := g.cl().CreateCacheParameterGroup(ctx, &elasticache.CreateCacheParameterGroupInput{
		CacheParameterGroupName:   aws.String(name),
		CacheParameterGroupFamily: aws.String("redis7"),
		Description:               aws.String("describe test"),
	})
	if err != nil {
		return err
	}
	resp, err := g.cl().DescribeCacheParameterGroups(ctx, &elasticache.DescribeCacheParameterGroupsInput{
		CacheParameterGroupName: aws.String(name),
	})
	if err != nil {
		return fmt.Errorf("elasticache DescribeCacheParameterGroups: describe failed: %w", err)
	}
	if len(resp.CacheParameterGroups) != 1 {
		return fmt.Errorf("elasticache DescribeCacheParameterGroups: expected 1 group, got %d", len(resp.CacheParameterGroups))
	}
	if aws.ToString(resp.CacheParameterGroups[0].CacheParameterGroupName) != name {
		return fmt.Errorf("elasticache DescribeCacheParameterGroups: expected CacheParameterGroupName %q, got %v", name, aws.ToString(resp.CacheParameterGroups[0].CacheParameterGroupName))
	}
	if aws.ToString(resp.CacheParameterGroups[0].CacheParameterGroupFamily) != "redis7" {
		return fmt.Errorf("elasticache DescribeCacheParameterGroups: expected CacheParameterGroupFamily 'redis7', got %v", aws.ToString(resp.CacheParameterGroups[0].CacheParameterGroupFamily))
	}
	return nil
}

func (g *elasticacheGroup) DeleteCacheParameterGroup(ctx context.Context, t *harness.TestContext) error {
	name := g.uniqueName("pg-del")
	t.Set("_pgDelName", name)
	_, err := g.cl().CreateCacheParameterGroup(ctx, &elasticache.CreateCacheParameterGroupInput{
		CacheParameterGroupName:   aws.String(name),
		CacheParameterGroupFamily: aws.String("redis7"),
		Description:               aws.String("delete test"),
	})
	if err != nil {
		return err
	}
	_, err = g.cl().DeleteCacheParameterGroup(ctx, &elasticache.DeleteCacheParameterGroupInput{
		CacheParameterGroupName: aws.String(name),
	})
	if err != nil {
		return fmt.Errorf("elasticache DeleteCacheParameterGroup: delete failed: %w", err)
	}
	_, err = g.cl().DescribeCacheParameterGroups(ctx, &elasticache.DescribeCacheParameterGroupsInput{
		CacheParameterGroupName: aws.String(name),
	})
	if err == nil {
		return fmt.Errorf("elasticache DeleteCacheParameterGroup: parameter group still exists after delete")
	}
	return nil
}

func (g *elasticacheGroup) teardownParameterGroups(ctx context.Context, t *harness.TestContext) error {
	for _, key := range []string{"_pgName", "_pgDescName", "_pgDelName"} {
		if name := t.GetString(key); name != "" {
			g.cl().DeleteCacheParameterGroup(ctx, &elasticache.DeleteCacheParameterGroupInput{CacheParameterGroupName: aws.String(name)}) //nolint:errcheck
		}
	}
	return nil
}
